cmd/hydrate/src/lib.rs

```diff
@@ -0,0 +1,217 @@
+    file: PathBuf,
+}
+
+struct DryCore {
+    flash: HubrisFlashMap,
+    mem: BTreeMap<u32, Vec<u8>>,
+}
+
+// Helper macro to stub out functions
+macro_rules! unsupported{
+    ($fn_name:ident($($arg_name:ident: $arg_type:ty),*)) => {
+        unsupported!($fn_name($($arg_name: $arg_type),*) -> Result<()>);
+    };
+    ($fn_name:ident($($arg_name:ident: $arg_type:ty),*) -> $out:ty) => {
+        fn $fn_name(&mut self, $($arg_name: $arg_type),*) -> $out {
+            bail!(concat!(
+                "DryCore does not support ",
+                stringify!($fn_name)))
+        }
+    };
+}
+
+impl humility::core::Core for DryCore {
+    unsupported!(run());
+    unsupported!(halt());
+    unsupported!(step());
+    unsupported!(load(_path: &std::path::Path));
+    unsupported!(reset());
+    unsupported!(write_8(_addr: u32, _data: &[u8]));
+    unsupported!(op_done());
+    unsupported!(op_start());
+    unsupported!(read_reg(_reg: ARMRegister) -> Result<u32>);
+    unsupported!(init_swv());
+    unsupported!(read_swv() -> Result<Vec<u8>>);
+    unsupported!(write_reg(_reg: ARMRegister, _value: u32));
+    unsupported!(write_word_32(_addr: u32, _data: u32));
+    unsupported!(reset_and_halt(_dur: std::time::Duration));
+    unsupported!(wait_for_halt(_dur: std::time::Duration));
+
+    fn info(&self) -> (String, Option<String>) {
+        ("DryCore".to_owned(), None)
+    }
+    fn is_archive(&self) -> bool {
+        false
+    }
+    fn is_dump(&self) -> bool {
+        true // I guess?
+    }
+    fn is_net(&self) -> bool {
+        false
+    }
+    fn vid_pid(&self) -> Option<(u16, u16)> {
+        None
+    }
+
+    fn read_8(&mut self, addr: u32, data: &mut [u8]) -> Result<()> {
+        if self.flash.read(addr, data).is_some() {
+            return Ok(());
+        }
+
+        let Some((base, mem)) = self.mem.range(0..=addr).next_back() else {
+            bail!("addr {addr:#08x} is below memory range");
+        };
+        let offset = (addr - base) as usize;
+        let end = offset + data.len();
+        if end > mem.len() {
+            bail!("region is not large enough; {end:#x} > {:#x}", mem.len());
+        }
+        data.copy_from_slice(&mem[offset..end]);
+        Ok(())
+    }
+}
+
+fn run(context: &mut ExecutionContext) -> Result<()> {
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+    let subargs = Args::try_parse_from(subargs)?;
+    let f = std::fs::File::open(&subargs.file)?;
+    let mut z = zip::ZipArchive::new(f)?;
+
+    let mut s = String::new();
+    z.by_name("meta.json")?.read_to_string(&mut s)?;
+    #[derive(serde::Deserialize)]
+    struct Meta {
+        version: u64,
+    }
+    let meta: Meta = serde_json::from_str(&s)?;
+    if meta.version != 1 {
+        bail!(
+            "invalid version in `meta.json`: expected 1, got {}",
+            meta.version
+        );
+    }
+
+    let mut s = String::new();
+    z.by_name("TASK_INDEX")?.read_to_string(&mut s)?;
```

Comment:
nit: looks like the `ZipError::FileNotFound` variant from the `zip` crate doesn't include the filename, so perhaps we ought to do something like:
```suggestion
    z.by_name("TASK_INDEX")
      .context("invalid dump: missing TASK_INDEX")?
      .read_to_string(&mut s)
      .context("failed to read TASK_INDEX from dump")?;
```
or similar, here and for every other time we read a file?

if that's too wordy, we could also have a helper like
```rust
fn read_zip_entry(z: &mut ZipArchive, file: &str) -> anyhow::Result<String> {
    let mut s = String::new();
    z.by_name(file)
      .with_context(|| format!("invalid dump: missing {file}"))?
      .read_to_string(&mut s)
      .with_context(|| format!("failed to read {file} from dump"))?;
    Ok(s)
}
```
and just call that everywhere we read something from the dry dump?

Reply:
This is somewhat moot because I moved metadata into the JSON file, but I added a few `with_context(...)` annotations elsewhere.